algorithms/src/snark/plookup/table.rs

```diff
@@ -0,0 +1,40 @@
+// Copyright (C) 2019-2022 Aleo Systems Inc.
+// This file is part of the snarkVM library.
+
+// The snarkVM library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkVM library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkVM library. If not, see <https://www.gnu.org/licenses/>.
+
+use snarkvm_fields::Field;
+
+pub struct LookupTable<F: Field>(Vec<Vec<F>>);
```

Comment:
For any circuit of interest, I believe that (at least) the number of columns is statically known. So I believe we can change this to `LookupTable<F: Field, const C: usize>(Vec<[F; C]>);`, which should reduce the number of redirections when accessing this vector.

Reply:
It makes sense - however, this might kind of get in the way when we want to have multiple lookup tables with different columns amounts. If that's no issue, then I'll opt for this but otherwise it may be better to stay unconstrained in this regard

Reply:
Although from an API point of view it is a bit nicer to constrain here since we won't have to deal with edge cases that come with needing to take in a dynamic-sized vector of lookup values - I'm also assuming we may get some performance gain out of being able to leverage arrays.

In order to save on space, I assume I can just manually implement the serialization-deserialization process to ignore zero elements in tables where the arity is smaller than the constrained size. It will sadly still be inefficient in-memory, but maybe we can find another way around this later.

Reply:
After quite a bit more experimentation, it seems that adding this const generic has an enormous cascading effect on the required generics for every single type related to calculating SNARKs, including the `SNARK` trait itself (as it uses `ConstraintSynthesizer`s which directly extend `ConstraintSystem`, which contains the `LookupTable<F, C>`).

If this is really desired, I can proceed on finalizing this work, but it has a bad design smell IME. It may be more convenient to instead opt for an `Indexmap` which will just allow map-like lookups, and constrain key lengths internally.